Add new keyboard shortcuts help modal

talldan opened the pull request:
## Description
Related issues - #71, #2980, #5420

Pressing `ctl+option+h` (`shift+alt+h` on windows/linux) displays a modal detailing keyboard shortcuts.

## How has this been tested?
- Manual testing (both Mac & Windows)
  - Test opening/closing modal using keyboard (esc also closes, as does hitting enter while the close 
button is focused)
  - Test opening/closing modal using mouse (open from more menu, close using button or clicking surrounding space)
  - Test classic block (that TinyMCE shortcut help doesn't also open at the same time as gutenberg shortcut help)
- Snapshot tests
- Added an e2e test that opens and closes modal using mouse and keyboard

## Screenshots
![screen shot 2018-08-09 at 4 14 10 pm](https://user-images.githubusercontent.com/677833/43886976-f4643d26-9bef-11e8-8f0e-7ba2b2c174c7.png)

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
New feature (non-breaking change which adds functionality) 
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

talldan committed:
Create a simple modal for showing keyboard shortcuts that displays when pressing cmd + /

talldan committed:
Add more shortcuts and place them in sections with a title and description

talldan committed:
Add styles for shortcut modal

talldan committed:
Add missing shortctuts

talldan commented:
Hey @jasmussen - before I take this one any further I thought I'd seek some design / copy feedback.

This has the same issue as the other PR with the formatting of shortcuts (e.g. ⌘+? instead of ⌘?), so I'll work on that now and cherry-pick the change onto both branches.

jasmussen commented:
Very very cool work, thank you for working on this.

I can't get the modal to show. `meta+?` isn't working for me. Also, can we make it `Shift+?` instead?  This keyboard shortcut is standard in Google properties and others, might as well reuse existing patterns if we can.

Finally, can we add it as a menu item? Like this:

![screen shot 2018-07-31 at 12 07 54](https://user-images.githubusercontent.com/1204802/43453419-62872760-94ba-11e8-9ac0-1bc768c52872.png)

I think this will be very welcome.

Just looking at the screenshot, I think this is a great start. I like how Google Inbox shows their shortcuts:

![screen shot 2018-07-31 at 12 10 59](https://user-images.githubusercontent.com/1204802/43453654-e795b764-94ba-11e8-9b70-9a6f79e1c959.png)

Note the little gray background behind each shortcut button. We can probably use the <kbd>kbd</kbd> element for this: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/kbd — possibly with a little minimalist styling.

Notably I like the relaxed row-like layout of keyboard shortcuts in Chrome OS:

![screen shot 2018-07-31 at 12 10 18](https://user-images.githubusercontent.com/1204802/43453738-1ea5b9b6-94bb-11e8-89ff-d80ff041fc46.png)

Flagging also @karmatosed for thoughts.

michelleweber commented:
(On the copy front, I'm not sure that you need the first sentence/subhead in each section; they bulk the modal up, but don't add any info that's not already there in the title and instructions. It also seems like the slash inserter could be part of the block shortcuts section, and I wonder if selection shortcuts can be part of global; then you have a simpler structure: general, block-specific, text-specific.)

karmatosed commented:
I would many +1's the way Google show this as being a good pattern for us to follow. Thanks for all your work on this @talldan.

talldan committed:
Remove unecessary shortcut section descriptions

talldan committed:
Display shortcut keys in a kbd element

talldan committed:
Styling changes

- Position shortcut to the right of description
- Adjust margin/padding
- Make section titles default font size

talldan committed:
Regroup shortcuts

talldan committed:
Fix invalid style rule

talldan committed:
Change copy for slash inserter shortcut

talldan committed:
Map over config to produce ShortcutSections

talldan committed:
Change where keyboard shortcut help modal is rendered

Move modal so that it's rendered in edit-post/layout component. This ensures
the modal is available in both the text and visual editor, and the shortcut
for launching the modal can be used on first load of the editor.

talldan commented:
@jasmussen I've made some changes today based on the feedback. Here's how it looks now:
![screen shot 2018-08-01 at 4 50 09 pm](https://user-images.githubusercontent.com/677833/43511556-46395834-95ab-11e8-9687-aac03abb6b77.png)

> I can't get the modal to show.
Looks like the editor needed focus for shortcut to work, I've fixed that now and it should work regardless.

> Also, can we make it Shift+? instead? 
The drawback with Shift+? is that it can't be used while typing, whereas Cmd+? can be. I'm happy to change it but thought I'd at least make you aware of why I went for Cmd+?.

> Finally, can we add it as a menu item
I still need to work on this. Will pick that up next.

jasmussen commented:
> The drawback with Shift+? is that it can't be used while typing, whereas Cmd+? can be. I'm happy to change it but thought I'd at least make you aware of why I went for Cmd+?.

Fair point. @karmatosed any thoughts? Go for Google consistency or go for the meta key? 

Might meta+? conflict with other things?

Looking forward to the menu item!

Modal looks great. Awesome work. A few small things:

1. "Select all" — might be cool to rephrase this so it notes that you can do multiple consecutive instances of this. For example

- Select all blocks when no block is selected, or select all text in current text block    ⌘A
- Select all blocks from within a text block    ⌘A ⌘A 

Something like that, but perhaps with @michelleweber help.

2. Typo: "Delete the select block(s)" — should be "selected" no?

On the ticket it was noted that ⌘+ duplicates for zoom, and ⌘D duplicates for bookmark. What other combos can we use here? ⌘+Shift+ + ? ⌘+Shift+D?

earnjam commented:
@talldan Design is looking good! I'm not a designer, but have two small points of feedback:
1. The section headers sort of blend in with the shortcut descriptions making it hard to find the section you're looking for. Perhaps a heavier weight, or larger size would help? 
2. It's a bit difficult to trace across the line to the actual keyboard shortcut from the shorter descriptions. There is a lot of scanning eyes back and forth to read down the list. 
Google handles this in the design posted above by putting the shortcut on the left side and aligning them right so they are flush with the descriptions. 
This is what Cheatsheet for OSX does as well:
![image](https://user-images.githubusercontent.com/4267290/43528771-bfc9e3d2-9577-11e8-88f0-4fb061dcab41.png)

Slack adds little grey lines to help easily align shortcut to description
![image](https://user-images.githubusercontent.com/4267290/43528665-7b2e5186-9577-11e8-801f-3ee3f7a2685f.png)

talldan committed:
Move keyboard-shortcut-help-modal to edit-post from editor package

talldan committed:
Add some border between rows in the keyboard shortcuts list

talldan committed:
Add reduxy things for opening and closing a modal

talldan committed:
Use actions to open/close modal and selector to determine current state

talldan committed:
Migrate the activeModal state from preferences to normal state so that its value is not persisted

talldan committed:
Add a menu item tot he more menu for toggling keyboard shortcut help

talldan committed:
Modify some shortcuts and description text

talldan committed:
Add snapshot tests for help modal

talldan committed:
Destructure props

talldan committed:
Modify header styling

talldan committed:
Rename class name to reflect component is now in edit-post, and reorder shortcuts

talldan commented:
Here's another update:
![screen shot 2018-08-02 at 5 45 22 pm](https://user-images.githubusercontent.com/677833/43576446-fef83766-967b-11e8-8266-897a6f31d627.png)

I think I've addressed a lot of the feedback
- Update copy and some of the shortcuts
- Added some border between the rows
- Add an option to the more menu to open the modal

@jasmussen - I haven't been able to find any issues with using Cmd+?, but I haven't checked Ctrl+? on Windows or Linux. I'll start looking at the best way to set up my windows box tomorrow.

talldan commented:
Just spotted, I should also add the shortcut to the close modal button :)

jasmussen commented:
Love it. This is really coming together. Thanks for adding the menu item, modal works great.

I still can't get ⌘+? to work on my danish keyboard. Question mark is the shift option of the plus button, so if I press ⌘? I zoom in. If I press ⌘+Shift+? I open Chrome help.

I think we either need to find a different shortcut — perhaps one localized to each keyboard, or one that leverages a key that's in the same place in all locales. Or, just be fine with NOT having a keyboard shortcut for help.

Does TinyMCE have a keyboard shortcut for help, that we can use?

Also I think it's worth adding that if you are in text, and press ⌘A you first select all the text in that block, but if you press ⌘A again (with that text selected), you expand to select all blocks. It's a cool feature and worth noting.

afercia commented:
> Does TinyMCE have a keyboard shortcut for help, that we can use?

It uses Ctrl+Alt+H

afercia commented:
I'm not sure I see some of the keyboard shortcuts specific for keyboard navigation, e.g.:

`'alt+f10': this.focusToolbar`

and the ones to jump through the main landmark regions:
```
'ctrl+`': this.focusNextRegion
'ctrl+shift+`': this.focusPreviousRegion
```

Worth noting https://github.com/WordPress/gutenberg/pull/8005 is going to add a couple additional shortcut.

I'd suggest to group al the shortcuts related to navigation under a group "Navigation".

afercia commented:
> I still can't get ⌘+? to work on my danish keyboard. Question mark is the shift option of the plus button, so if I press ⌘? I zoom in.

Can confirm it doesn't work with an Italian keyboard layout too. Worth noting using `?` for the shortcut exposes to the same problem faced for other shortcuts: not all the keyboard layout have non-alphanumeric keys in the same location: https://it.wikipedia.org/wiki/File:Italian_Keyboard_layout.svg

<img width="813" alt="screen shot 2018-08-02 at 14 37 14" src="https://user-images.githubusercontent.com/1682452/43584067-9b57ddd4-9661-11e8-85a5-be6e4172d4de.png">

Tested on Window Chrome:
- pressing Ctrl + single quote or Ctrl + Shift + single quote doesn't do anything
- what works, surprisingly, is Shift + ù

Given all the exploration made in #3218 I'd suggest to use two modifiers and a character-key.

chrisvanpatten commented:
> The drawback with Shift+? is that it can't be used while typing, whereas Cmd+? can be. I'm happy to change it but thought I'd at least make you aware of why I went for Cmd+?.

FWIW, here are what a few popular web apps do…

+ Gmail: <kbd>shift+/</kbd>
+ Slack: <kbd>Cmd+/</kbd>
+ Google Docs: <kbd>Cmd+/</kbd>
+ GitHub: <kbd>Cmd+/</kbd>

afercia commented:
No, also forward slash is placed differently in various keyboard layouts. See the link to the wikipedia page in https://github.com/WordPress/gutenberg/issues/3218#issuecomment-409135479

talldan commented:
I spent too much time looking into this today, and have come to the conclusion that I don't think there's an easy way to bind either `cmd/ctrl + ?` or `cmd/ctrl + /` and have it work well across keyboard layouts.

Looking at how slack handles it, they use a different shortcut depending on the keyboard layout:
https://get.slack.help/hc/en-us/articles/115005429666-Manage-your-keyboard-layout

I tested and in Italian and Danish layouts they use `cmd + .`.

Google docs doesn't bother handling this at all.

We could consider using `cmd+k` as the shortcut. Either on its own or as an additional shortcut.

`?` on its own may be possible, though it'd not be possible while typing in a block.

talldan committed:
Display shortcut for more menu menu item

jasmussen commented:
I personally think it's okay if the help shortcut doesn't work while typing. In fact I also think it's okay to not have a shortcut at all — you can access the More menu using the keyboard and then invoke the help modal from there. Perhaps in order to ship the improvement that is this modal in the first place, it would be good to remove the shortcut altogether and possibly revisit later?

Did you try `Ctrl+Alt+H` which is what TinyMCE uses? Be sure to test also in the classic block.

noisysocks commented:
Is it easy to detect the keyboard layout and do what Slack does? Otherwise I reckon let's borrow the TinyMCE shortcut for now.

jasmussen commented:
> Is it easy to detect the keyboard layout and do what Slack does? 

I can't say with certainty because I'm not as skilled codewise as many of you are. And I would also love us to do this — it feels like the way superior solution. But my inkling is that it is extremely difficult, very onerous and time consuning, and prone to error as there are many many keyboard layouts we have to account for. Given how close we are to 5.0 deadlines, I'm thinking we should either keep things simple in order to ship improvements and iterate later, or just ship the improvements and be fine with not having any, or a clever, shortcut.

talldan committed:
Disable WP_Help TinyMCE command for classic block

talldan committed:
Change keyboard shortcut for triggering help to access + h, the same as the classic editor

talldan commented:
@noisysocks @jasmussen I don't think it's possible given the limited information the web exposes in keyboard events. Slack get around it by binding both keys.

Anyway - I've had some success, Ctrl+Alt+H seems to work well and I've disabled the classic block's help modal in this commit - 9521a07

talldan committed:
Add utility to keycodes for returning key combinations in an array

talldan committed:
Avoid ugly regular expression to split key combination into array

Using newly introduced displayShortcutList function results in cleaner, better tested code

talldan commented:
Things remaining:
- Update some shortcuts
- Add any missing shortcuts
- Cross-browser/os test
- Code review
- Code review fixes
- Code review
- Code review fixes
- Code review
- Code review fixes

:)

afercia commented:
Noting that https://github.com/WordPress/gutenberg/pull/8005 is going to add two alternative keyboard shortcuts for the navigate regions component, dedicated to keyboard layouts without the backtick key.

karmatosed reviewed (approved):
I really like this! Excited to get it. Just as a note, lets once this is in loop around to getting help itself (not just shortcuts) in as a link using the same modal (maybe with tabs).

noisysocks commented on core-blocks/freeform/edit.js:
```diff
@@ -84,6 +84,13 @@ export default class FreeformEdit extends Component {
+			if ( event.command === 'WP_Help' ) {
+				event.preventDefault();
+			}
+		} );
```
Wondering if `editor.shortcuts.remove( 'ctrl+alt+h' )` might be a nicer way of doing this? Not sure.

https://www.tiny.cloud/docs/api/tinymce/tinymce.shortcuts/

talldan replied:
Doesn't seem to work unfortunately. I also thought `editor.editorCommands.addCommand( 'WP_Help', noop  )` might work but it doesn't. It suggests a race condition - that the command is added after freeforms setupEditor runs.

noisysocks replied:
Does what we're doing in RichText help?

https://github.com/WordPress/gutenberg/blob/4a78caed70cf899fa7bcd42258adbcafe5d5dda7/packages/editor/src/components/rich-text/index.js#L169

talldan replied:
Sadly not. Guessing it works there because Redo is already bound by TinyMCE at this point.

noisysocks commented on edit-post/components/header/keyboard-shortcuts-toggle/index.js:
```diff
@@ -0,0 +1,43 @@
+			onClick={ toggleModal }
+			shortcut={ displayShortcut.access( 'h' ) }
+		>
+			{ __( 'Keyboard Shortcuts' ) }
```
Looks a little strange that this copy wraps onto two lines. Could we widen the menu or shorten the text?

Also, the shortcut text on the right isn't vertically centred.

<img width="277" alt="screen shot 2018-08-09 at 12 10 17" src="https://user-images.githubusercontent.com/612155/43874101-36315e14-9bcd-11e8-93f8-c7cf9d2fbc89.png">

talldan replied:
Yeah, I spotted that. I wonder if shortening Option to Opt might be something we'd consider. Or using the symbol. We might still have this issue with translations, so worth making it look as nice as possible when wrapped.

@jasmussen any opinions on this?

noisysocks replied:
> We might still have this issue with translations.

Yeah, let's definitely vertically centre the text on the right regardless of what we do here.

talldan replied:
This is what it looks like with the option symbol:
![screen shot 2018-08-09 at 3 35 15 pm](https://user-images.githubusercontent.com/677833/43884887-e989a874-9be9-11e8-8dd7-94f56b351863.png)

talldan replied:
If we wanted to go crazy, could also use ⇧ for shift.

noisysocks replied:
It looks like https://github.com/WordPress/gutenberg/pull/8756 might solve this 🙂

noisysocks commented on edit-post/components/header/keyboard-shortcuts-toggle/index.js:
```diff
@@ -0,0 +1,43 @@
+
+export default compose( [
+	withSelect( ( select ) => ( {
+		isModalActive: select( 'core/edit-post' ).isModalActive( modalName ),
```
I'm thinking it might make more sense to make this a regular button that, when clicked, opens the modal and closes the dropdown menu (using a passed in `onClose` prop). 

This would also fix some weirdness that exists right now where the dropdown menu is dismissed when you click on the shortcuts modal.

![menu](https://user-images.githubusercontent.com/612155/43874286-580a32f8-9bce-11e8-80ad-64ed83d37028.gif)

talldan replied:
Oh wow - good catch!

noisysocks commented on edit-post/components/header/keyboard-shortcuts-toggle/index.js:
```diff
@@ -0,0 +1,43 @@
+import { compose } from '@wordpress/compose';
+import { MenuItem } from '@wordpress/components';
+
+const modalName = 'edit-post/keyboard-shortcut-help';
```
We generally use SCREAMING_SNAKE_CASE for constants 😱🐍

noisysocks commented on edit-post/components/keyboard-shortcut-help-modal/config.js:
```diff
@@ -0,0 +1,115 @@
+/**
```
Having this all in a `config.js` is a nice idea 🙂🌈

noisysocks commented on edit-post/components/keyboard-shortcut-help-modal/index.js:
```diff
@@ -0,0 +1,119 @@
+import shortcutConfig from './config';
+import './style.scss';
+
+const modalName = 'edit-post/keyboard-shortcut-help';
```
😱🐍

noisysocks commented on edit-post/components/keyboard-shortcut-help-modal/index.js:
```diff
@@ -0,0 +1,119 @@
+			>
+				<dt className="edit-post-keyboard-shortcut-help__shortcut-term">
+					<kbd className="edit-post-keyboard-shortcut-help__shortcut-key-combination">
+						{ mapKeyCombination( isArray( keyCombination ) ? keyCombination : [ keyCombination ] ) }
```
You could use `_.castArray` here.

```js
{ mapKeyCombination( castArray( keyCombination ) ) }
```

noisysocks commented on edit-post/components/keyboard-shortcut-help-modal/style.scss:
```diff
@@ -0,0 +1,56 @@
+	}
+
+	&__shortcut-key {
+		padding: .25rem .5rem;
```
If you pull in master you'll now get some `stylelint` warnings here 🙂

noisysocks commented on edit-post/store/reducer.js:
```diff
@@ -88,6 +88,17 @@ export function panel( state = 'document', action ) {
 	return state;
 }
 
+export function activeModal( state = null, action ) {
```
Can you please add a JSDoc comment describing what this reducer is for?

noisysocks commented on edit-post/store/actions.js:
```diff
@@ -22,6 +22,30 @@ export function closeGeneralSidebar() {
+ * Returns an action object used in signalling that the user opened an editor sidebar.
+ *
+ * @param {string} name        Sidebar name to be opened.
+ * @return {Object}            Action object.
```
Indentation here is a bit off. I think all of the fields are supposed to line up, and `@return` should have a blank line above it.

(I know, I know, we really need lint rules for this...)
```
 * @param {string} name Sidebar name to be opened.

 * @return {Object} Action object.
```

noisysocks commented on edit-post/store/selectors.js:
```diff
@@ -96,6 +96,17 @@ export function isEditorSidebarPanelOpened( state, panel ) {
+ *
+ * @param  {Object}  state Global application state.
+ * @param  {string}  modalName Modal name.
+ * @return {boolean}       Whether the modal is active.
```
```
 * @param {Object} state     Global application state.
 * @param {string} modalName Modal name.
 *
 * @return {boolean} Whether the modal is active.
```

Also, `Modal name.` isn't very descriptive. Maybe something like `A string that uniquely identifies the modal.`?

noisysocks commented on edit-post/store/test/reducer.js:
```diff
@@ -153,6 +154,23 @@ describe( 'state', () => {
 		} );
 	} );
 
+	it( 'should set the activeModal to the provided name', () => {
```
Can you please group these tests into a `describe( 'activeModal', () => { ... } )`?

noisysocks commented on edit-post/store/test/reducer.js:
```diff
@@ -153,6 +154,23 @@ describe( 'state', () => {
+		} );
+
+		expect( state ).toBeNull();
+	} );
```
We should add a test that verifies the initial state.

```js
	it( 'should default to null', () => {
		const state = activeModal( undefined, {} );
		expect( state ).toBeNull();
	} );
```

noisysocks commented on edit-post/store/test/selectors.js:
```diff
@@ -125,6 +126,34 @@ describe( 'selectors', () => {
+				preferences: {
+					activeModal: null,
+				},
+			};
```
We shouldn't have `preferences` here.

```js
const state = {
	activeModal: null,
};
```

noisysocks reviewed (changes requested):
Very cool! ✨🌈

I left a few minor comments.

talldan committed:
Add e2e test for shortcut help modal

talldan committed:
Vertically center shortcut text (if adjacent label wraps onto multiple lines)

talldan committed:
Close more menu when keyboard shortcuts modal is opened using the menu item

talldan committed:
Use castArray over lengthy ternary

talldan committed:
Styling linter fixes

talldan committed:
JSDoc fixes

talldan committed:
Group activeModal unit tests in a describe

talldan committed:
Add test for initial activeModal state

talldan committed:
Fix incorrect state used in test

talldan committed:
Update list of key shortcuts for the help modal

noisysocks commented on test/e2e/support/utils.js:
```diff
@@ -205,19 +214,23 @@ export async function insertBlock( searchTerm, panelName = null ) {
 	await page.keyboard.press( key );
-	return page.keyboard.up( modifier );
+
+	return await Promise.all(
```
We don't need the `return` here since we're `await`ing the promise.

noisysocks commented on edit-post/components/header/keyboard-shortcuts-help-menu-item/index.js:
```diff
@@ -0,0 +1,46 @@
+
+export default compose( [
+	withSelect( ( select ) => ( {
+		isModalActive: select( 'core/edit-post' ).isModalActive( MODAL_NAME ),
```
Do we need this? When would you be able to close the modal using this button?

talldan replied:
very true - removed the toggle functionality, it's now just an opener.

talldan committed:
Remove unecessary return

talldan committed:
Make KeyboardShortcutsHelpMenuItem only open the help modal

noisysocks reviewed (approved):
Love it! Great job.

aduth commented on test/e2e/support/utils.js:
```diff
@@ -235,19 +244,21 @@ export async function insertBlock( searchTerm, panelName = null ) {
+	const modifierKeys = castArray( modifiers );
+
+	await Promise.all(
+		modifierKeys.map( async ( modifier ) => page.keyboard.down( modifier ) )
```
Minor: There's no need for this to be an `async` function, since the `await` keyword is not used within the function's body.
